src/cpp/lpnamer/input_reader/WeightsFileWriter.cpp

```diff
@@ -11,34 +13,35 @@ YearlyWeightsWriter::YearlyWeightsWriter(
     const std::filesystem::path& xpansion_output_dir,
     const std::vector<double>& weights_vector,
     const std::filesystem::path& output_file,
-    const std::vector<int>& active_years)
+    const std::vector<int>& active_years, const std::string& solver_name)
     : xpansion_output_dir_(xpansion_output_dir),
       weights_vector_(weights_vector),
       output_file_(output_file),
-      active_years_(active_years) {
+      active_years_(active_years),
+      solver_name_(solver_name) {
   xpansion_lp_dir_ = xpansion_output_dir / LP_DIR;
   if (!std::filesystem::is_directory(xpansion_lp_dir_)) {
     std::filesystem::create_directory(xpansion_lp_dir_);
   }
 }
 
-void YearlyWeightsWriter::CreateWeightFile() {
-  FillMpsWeightsMap();
+void YearlyWeightsWriter::CreateWeightFile(
+    const std::vector<std::pair<std::shared_ptr<Problem>, ProblemData>>&
+        problems_and_data) {
+  FillMpsWeightsMap(problems_and_data);
   DumpMpsWeightsToFile();
 }
 
-void YearlyWeightsWriter::FillMpsWeightsMap() {
+void YearlyWeightsWriter::FillMpsWeightsMap(
+    const std::vector<std::pair<std::shared_ptr<Problem>, ProblemData>>&
+        problems_and_data) {
   mps_weights_.clear();
-  auto it = std::filesystem::directory_iterator(xpansion_lp_dir_);
-  auto mps_list = it | std::views::filter([](const auto& p) {
-                    return p.path().extension() == ".mps";
-                  });
-  for (auto& mps_file : mps_list ) {
-    auto year = GetYearFromMpsName(mps_file.path().filename().string());
-    auto year_index =
-        std::find(active_years_.begin(), active_years_.end(), year) -
-        active_years_.begin();
-    mps_weights_[mps_file.path().filename()] = weights_vector_[year_index];
+
+  for (const auto& [problem, data] : problems_and_data) {
+    auto year_index = std::find(active_years_.begin(), active_years_.end(),
+                                problem->mc_year) -
+                      active_years_.begin();
```

Comment:
Returns an iterator to the first element in the range [first, last) that satisfies specific criteria (or last if there is no such iterator). 

`std::find` could fail to find a matching element, in which case it would return `active_years_.end()`. This case is not handled and lead to an invalid `year_index`

Reply:
Also, you convert a position in a vector to a year. This isn't very robust since the order could change in the future, there could be missing elements, etc.

Reply:
No `year_index` is not a year, it is a position in a vector... this line already exists in the code, it is not new by the way (even though I agree it can be more robust)

Reply:
Existing `problem->mc_year` are exactly the active years of the study, and we need to retrieve the index of the current mc year in the list of active years to read the correct line of the weights file. I do not know how we can do this much better (except by changing the data structure of `active_years` to store a tuple (year_index, year) which may be overkill)

Reply:
Log message added in case `mc_year` not found in `active_years`